docs/proposals/odo-v3-cli/odo-delete-component.md

```diff
@@ -0,0 +1,17 @@
+# `odo delete component`
+Delete component.
+By default it deletes component from cluster (deletes all resources from both `odo dev` and `odo deploy` that were crated on the cluster that belong to the component) that in the current directory.
+
+Similarly to `odo describe component` this commands works in two modes.
+Without `--component` `--application` or `--namespace` it works with the component in the current directory.
+If user defines `--component` and `--application` than it works with the component that should exist on the cluster, and it should not touch local component even if the current directory is component directory.
+`--namespace` flag is optional, but it makes sense only together with `--component` and `--application`, if not used than it uses the current namespace as defined in `KUBECONFIG`.
+
+
+## flags
+- `-o json` - show output in json format
+- `-f` `--force` - force deletion, don't ask for confirmation
+- `--component` - name of the existing component on the cluster
+- `--application` - name of the application that the component belongs to
+- `--namespace` - namespace
```

Comment:
I request removing all these flags and make a user `cd` into the directory of the component they want to delete. As for IDE plugins, they can use `odo list` to get the path of the component and `cd` into it before doing `odo delete`. So, for IDE plugins, deletion would be a two-step thing - first `cd` and then `odo delete`. However, it could be invisible to the user.

Reply:
> I request removing all these flags and make a user cd into the directory of the component they want to delete. As for IDE plugins, they can use odo list to get the path of the component and cd into it before doing odo delete

That wouldn't work. The user needs to be able to delete components for which there is no local devfile.
What if I accidentally delete a component locally before deleting it from the cluster?

Reply:
> The user needs to be able to delete components for which there is no local devfile.
> What if I accidentally delete a component locally before deleting it from the cluster?

This, IMO, is an over-thinking scenario. Have we come across users who have faced this problem? As far as IDE users are concerned, they are interacting via IDE, so they are not going to fall into this issue. And for CLI, if a user does `rm -rf`, they can very well do `kubectl delete deployment` as well. If they don't know of this kubectl command, they can open an issue on repo, and we can guide them. Or a support ticket if that's what they would prefer.

IMO, we should focus on 80 part of the 80-20 principle - what matters to 80% of the users and do that.

Reply:
>This, IMO, is an over-thinking scenario. Have we come across users who have faced this problem?

Yes you are talking to one.

>. And for CLI, if a user does rm -rf, they can very well do kubectl delete deployment

It is not as easy as doing `kubectl delete deployment` odo component will typically create a lot more resources.

> IMO, we should focus on 80 part of the 80-20 principle - what matters to 80% of the users and do that.

This functionality is basic requirement that needs to be there.

Reply:
> Yes you are talking to one.

Ah, that's cool. I didn't know that you used odo for creating microservices using Spring/NodeJS/Postgres.

> It is not as easy as doing `kubectl delete deployment` odo component will typically create a lot more resources.

odo makes sure to tie everything to a deployment by using `ownerReferences`. So, unless I am mistaken, it is actually as simple as doing `kubectl delete deployment`.

> This functionality is basic requirement that needs to be there.

Disagree. But https://github.com/redhat-developer/odo/pull/5341#discussion_r796302382 should help with it. My request is to not bloat the commands with flags just because we can do it. Instead, add flags consciously when a functionality can't be achieved otherwise. I don't agree in doing this because the IDE plugins need it, either. They can easily execute a task by running two commands instead of one, no?

Reply:
>> It is not as easy as doing kubectl delete deployment odo component will typically create a lot more resources.

> odo makes sure to tie everything to a deployment by using ownerReferences. So, unless I am mistaken, it is actually as simple as doing kubectl delete deployment.

Yes right now it does. But as we discussed many times this is not a good design that needs to change, for example, we want to support "service-only" components in the future.

Reply:
The flags `--namespace`, `--context`, and similar are flags that modify the "context" of the application. 
One solution would be to let the user do these changes of context itself (cd $CONTEXT for `--context`, kubectl ns $NAMESPACE for `--namespace`, etc).

If we want to efficiently integrate these flags in odo, we should consider them just for this change of context: instead of polluting the code with passing the values of these flags, we should instead change the context as soon as possible in the flow of the code (probably long before we enter the code of a specific command), and forget them for the code we are working on in a daily basis.

As an example, most of the code can consider odo works on local directory, and you can change the current working directory at the very beginning (and don't forget to change it back before to quit).

Reply:
> for example, we want to support "service-only" components in the future.

I wager it'll require doing a v4, because in v3 our main focus is components. For example, `odo init` bootstraps a component. Component, based on what I understand, is the most basic unit of odo.

Reply:
`odo init` doesn't need to know how to create "service-only" components. `odo dev` will just need to understand how to create resources out of it,  `odo list/describe/delete` will also have to understand it.

We know about this requirement and we are building v3, so we need to consider this so it can be incorporated in the future without requiring v4.

Reply:
If we were to make `odo dev` ready for this, it should create cluster resources for the components that are not of type `image`. And then it should execute the commands for these components. But there has to be some order in which the commands are run, right? Does API define that? 

What else should I be thinking from devfile point of view?

Reply:
>If we were to make odo dev ready for this, it should create cluster resources for the components that are not of type image. And then it should execute the commands for these components. But there has to be some order in which the commands are run, right? Does API define that?

yes

Reply:
Anything else you think we should be keeping in mind? I'm asking this as someone who finds it sorta difficult to figure things out from devfile API.